### PR TITLE
Introduce "Ingested Entities"

### PR DESCRIPTION
**Pull request by LindvedKrvang**
SofieServer now has the concept of "Ingested Entities". These entities are the data that has been fetched from the IngestGateway.
SofieServer then copies these ingested entities over to its 'regular' entities and uses the 'regular' entities for everything else.
This gives us the benefit that we can do whatever we want with the entities and when we need to "reset" the entities, we can just "create a new one" from the corresponding ingested entity. For example, we don't need to keep track of unplanned Pieces on a Part anymore. We can just remove all Pieces that is not included in the `IngestedPieces` for the corresponding Part.

In essence, we now have a "source of ingested truth" and "current state of the Rundown".

 
Included in this PR:
- New "ingested entities" `IngestedRundown`, `IngestedSegment`, `IngestedPart` and `IngestedPieces`. With corresponding repositories and `MongoIngestedX` interfaces.
- A new `IngestedEntityConverter` to converter between Ingested Entities and our 'regular' entities.
- A streamlining of our 'regular' `MongoX` interfaces since we are no longer required to have the same naming convention in the database as Core has.
- SofieServer listens for Mongo events on the Ingested entities and updates the 'regular' entities.
- When SofieServer starts up, it checks in the Ingested entities collections and `synchronize` the 'regular' entities (taking into account Active Rundowns etc.)

### CODE REVIEW

File: src/model/entities/part.ts [hunk 1]
```diff
@@ -302,13 +317,30 @@ export class Part {
...
     this.pieces = [
       ...this.pieces.filter(piece => piece.isPlanned),
       ...this.replacedPlannedPieces
-    ]
+    ].filter(piece => this.defaultPart!.ingestedPieces.some(ingestPiece => ingestPiece.id === piece.id))
+      .map(piece => {
+        const ingestedPiece: IngestedPiece = this.defaultPart!.ingestedPieces.find(ingestPiece => ingestPiece.id === piece.id)!
+        piece.resetFromIngestedPiece(ingestedPiece)
+        return piece
+      })
```
Comment (andr9528) [inline src/model/entities/part.ts:343]: I think this has a size now that i would benefit from its own aptly named method. The assignment to one attribute has grow to 10 lines of code. At a minimum, the function of the map, should be moved out. As long as that is done, i am happy.

File: src/business-logic/services/ingested-entity-to-entity-mapper.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      segments: [],
+      history: [],
+      isRundownActive: false,
+      modifiedAt: ingestedRundown.modifiedAt,
+      baselineTimelineObjects: ingestedRundown.baselineTimelineObjects,
+      timing: ingestedRundown.timings
+    })
+  }
+
+  public updateRundownFromIngestedRundown(rundownToUpdate: Rundown, ingestedRundown: IngestedRundown): Rundown {
```
[the remark this answers is not in the document]
Comment (andr9528) [reply src/business-logic/services/ingested-entity-to-entity-mapper.ts:31]: The same goes for some of the other methods in this class.

**Comment by andr9528** [comment]
Have been through all the files and not found any additional things to comment one, apart from my other comments.

**Review by andr9528 — COMMENTED** [review]
Have been through all the files and not found any additional things to comment one, apart from my other comments.

File: src/business-logic/services/ingested-entity-to-entity-mapper.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/ingested-entity-to-entity-mapper.ts:31]: That's intentional. From a consumer, calling this class, point of view we just want to update our entity, so we call an update method.
Because of immutability we have to create a new instance.

File: src/model/entities/part.ts (hunk 1 above)
Comment (LindvedKrvang) [reply src/model/entities/part.ts]: I have extracted it to a method

File: src/business-logic/services/ingested-entity-to-entity-mapper.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (andr9528) [reply src/business-logic/services/ingested-entity-to-entity-mapper.ts:31]: Fair enough.

**Review by andr9528 — APPROVED** [review]

File: src/model/entities/part.ts [hunk 3]
```diff
@@ -31,10 +34,13 @@ export interface PartInterface {
 
   timings?: PartTimings
   endState?: PartEndState
+
+  defaultPart?: IngestedPart
```
Comment (RasmusAlbrektsen) [inline src/model/entities/part.ts:38]: I'm not quite sure i understand, what is the responsibility of the `defaultPart`?

**Review by RasmusAlbrektsen — APPROVED** [review]
Looks good, i mostly just have some questions :-)

File: src/model/entities/part.ts (hunk 3 above)
Comment (LindvedKrvang) [reply src/model/entities/part.ts:38]: Maybe the naming isn't optimal.

It's the "ingested/default" Part. It's the values to fall back on when resetting the Part.
So we can change the values of the Part to our hearts desire. Then when we reset the Part, we change all the values back to the values from the `defaultPart`.

We are changing the values back on the Part instead of creating a new Part object instance so we can keep the same object references that is being used by Rundown and Segment.
Also, by letting the Part have its own default Part, we don't need to worry about mappers between `IngestedPart` and `Part` which is something we don't have access to within our entities.

File: src/data-access/repositories/cache/cached-part-repository.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  private static instance: PartRepository
+
+  public static getInstance(partRepository: PartRepository): PartRepository {
+    if (!this.instance) {
+      this.instance = new CachedPartRepository(partRepository)
+    }
+    return this.instance
+  }
+
+  private readonly cachedParts: Map<string, Part> = new Map()
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/cache/cached-part-repository.ts:14]: I don't think we have an official agreed upon preferred ordering.
What we have beeing following so far is all instance variables going above the constructor/methods. `cachedParts` is an instance variable and hence it is above the constructor.

My own ordering is:
- public static instance variables
- private static instance variables
- Depending on the case: static methods
- public instance variables
- private instance variables
- constructor
- methods

File: src/data-access/repositories/mongo/mongo-ad-lib-actions-repository.ts [hunk 5]
```diff
@@ -13,8 +12,8 @@ interface AdLibAction {
 }
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ad-lib-actions-repository.ts:12]: Newline added

File: src/business-logic/services/database-change-ingest-service.ts [hunk 6]
```diff
@@ -40,46 +55,170 @@ export class DatabaseChangeIngestService implements IngestChangeService {
...
 
-  private listenForSegmentChanges(segmentChangedListener: DataChangedListener<Segment>): void {
-    segmentChangedListener.onCreated(segment => this.enqueueEvent(() => this.createSegment(segment)))
-    segmentChangedListener.onUpdated(segment => this.enqueueEvent(() => this.updatedSegment(segment)))
-    segmentChangedListener.onDeleted(segmentId => this.enqueueEvent(() => this.deleteSegment(segmentId)))
+  private async syncDataWithIngestedData(): Promise<void> {
+    const ingestedRundowns: IngestedRundown[] = await this.ingestedRundownRepository.getIngestedRundowns()
+    await this.deleteRundownsNotPresentInIngestedRundowns(ingestedRundowns)
+
+    await Promise.all(ingestedRundowns.map(async (ingestedRundown) => {
```
Comment (KvelaGorrrrnio) [inline src/business-logic/services/database-change-ingest-service.ts:107]: The lambda function could be extracted to a new method e.g. `synchronizeIngestedRundown`.

File: src/business-logic/services/database-change-ingest-service.ts [hunk 7]
```diff
@@ -40,46 +55,170 @@ export class DatabaseChangeIngestService implements IngestChangeService {
...
+  private updateActiveRundownFromIngestedRundown(ingestedRundown: IngestedRundown, rundown: Rundown): Rundown {
+    const ingestedSegmentIds: string[] = ingestedRundown.ingestedSegments.map(ingestedSegment => ingestedSegment.id)
+    const segmentsToBeDeleted: Segment[] = rundown.getSegments().filter(segment => !ingestedSegmentIds.includes(segment.id))
+    segmentsToBeDeleted.forEach(segment => rundown.removeSegment(segment.id))
+
+
+    ingestedRundown.ingestedSegments.forEach(ingestedSegment => {
+      const segmentOnRundown: Segment | undefined = rundown.getSegments().find(segment => segment.id === ingestedSegment.id)
+      if (!segmentOnRundown) {
+        const newSegment: Segment = this.ingestedEntityToEntityMapper.fromIngestedSegment(ingestedSegment)
```
Comment (KvelaGorrrrnio) [inline src/business-logic/services/database-change-ingest-service.ts:152]: Is `new` necessary?

File: src/business-logic/services/database-change-ingest-service.ts [hunk 8]
```diff
@@ -40,46 +55,170 @@ export class DatabaseChangeIngestService implements IngestChangeService {
     return this.instance
   }
 
-  private readonly eventQueue: (() => Promise<void>)[] = []
+  private readonly eventPriorityQueue: Record<number, (() => Promise<void>)[]> = { }
```
Comment (KvelaGorrrrnio) [inline src/business-logic/services/database-change-ingest-service.ts:58]: It would be nice to know if the priorities of the enqueued elements are ascending or descending.

File: src/business-logic/services/database-change-ingest-service.ts [hunk 9]
```diff
@@ -40,46 +55,170 @@ export class DatabaseChangeIngestService implements IngestChangeService {
...
+    const rundownIds: string[] = basicRundowns.map(basicRundown => basicRundown.id)
+
+    for (const rundownId of rundownIds) {
+      // Delete all Rundowns that has been deleted and therefore no longer exists in the IngestedRundowns.
+      if (!ingestedRundownsIds.includes(rundownId)) {
+        await this.rundownRepository.deleteRundown(rundownId)
+        this.eventEmitter.emitRundownDeleted(rundownId)
+      }
+    }
+  }
```
Comment (KvelaGorrrrnio) [inline src/business-logic/services/database-change-ingest-service.ts:141]: Mapping basic rundowns to ids seems unnecessary. Here is a suggestion:
```typescript
  private async deleteRundownsNotPresentInIngestedRundowns(ingestedRundowns: IngestedRundown[]): Promise<void> {
    const ingestedRundownsIds: string[] = ingestedRundowns.map(ingestedRundown => ingestedRundown.id)
    const basicRundowns: BasicRundown[] = await this.rundownRepository.getBasicRundowns()
    
    for (const basicRundown of basicRundowns) {
      if (ingestedRundownsIds.includes(basicRundown.id) {
        continue
      }
      await this.rundownRepository.deleteRundown(basicRundown.id)
      this.eventEmitter.emitRundownDeleted(basicRundown.id)
    }
  }
```

I have removed the comment in the suggestion, since it says the same as the method name.

File: src/business-logic/services/database-change-ingest-service.ts [hunk 10]
```diff
@@ -104,24 +243,40 @@ export class DatabaseChangeIngestService implements IngestChangeService {
       })
   }
 
+  private getEventToExecute(): (() => Promise<void>) | undefined {
+    const events: (() => Promise<void>)[] | undefined = Object.entries(this.eventPriorityQueue)
+      .sort(([a], [b]) => Number.parseInt(a) - Number.parseInt(b))
```
Comment (KvelaGorrrrnio) [inline src/business-logic/services/database-change-ingest-service.ts:248]: `a` and `b` can be renamed to `priorityA` and `priorityB` to indicate that it is the priority that is sorted by.

You can consider using a Map instead of a record. That way priorities will keep their number type, when retrieving the entries.

Another solution will be to implement/use a max heap – but that is outside the scope of this PR :D

**Review by KvelaGorrrrnio — COMMENTED** [review]
Nice work :D

File: src/data-access/repositories/test/mongo-part-repository.spec.ts [hunk 11]
```diff
@@ -1,16 +1,22 @@
...
+import { MongoEntityConverter, MongoId, MongoPart } from '../mongo/mongo-entity-converter'
 
-const COLLECTION_NAME = 'parts'
+const COLLECTION_NAME: string = PART_COLLECTION_NAME
+
+describe('', () => {
+  it('', () => {
+    // Necessary test to circumvent that the tests are run as an exported function
+  })
+})
```
[the remark this answers is not in the document]
Comment (andr9528) [reply src/data-access/repositories/test/mongo-part-repository.spec.ts:19]: That is the explanation he gave to me atleast.

File: src/model/entities/part.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/part.ts]: Updated to make `isPlanned` a computed property.
Renamed `defaultPart` to `ingestedPart`

File: src/model/entities/part.ts [hunk 12]
```diff
@@ -69,8 +75,15 @@ export class Part {
    */
   private endState?: PartEndState
 
+  public readonly defaultPart?: IngestedPart
+
   constructor(part: PartInterface) {
+    if (part.isPlanned && !part.defaultPart) {
+      throw new UnsupportedOperation(`Found planned Part ${part.id} without a default Part. Planned Parts must have default Parts!`)
+    }
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/part.ts]: This has been deleted since `isPlanned` is now a computed property

File: src/data-access/repositories/cache/cached-part-repository.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  private deletePartsWithPredicate(predicate: (part: Part) => boolean): void {
+    const segmentIdsToDelete: string[] = []
+    this.cachedParts.forEach((part, partId) => {
+      if (predicate(part)) {
+        segmentIdsToDelete.push(partId)
+      }
+    })
+    segmentIdsToDelete.forEach(id => this.cachedParts.delete(id))
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/cache/cached-part-repository.ts]: Updated for Cached Segment and Part repositories

File: src/business-logic/services/ingested-entity-to-entity-mapper.ts [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+import { ErrorCode } from '../../model/enums/error-code'
+import { IngestedSegment } from '../../model/entities/ingested-segment'
+import { Segment } from '../../model/entities/segment'
+import { IngestedRundown } from '../../model/entities/ingested-rundown'
+import { Rundown, RundownAlreadyActiveProperties } from '../../model/entities/rundown'
+
+export class IngestedEntityToEntityMapper {
+
+  constructor() { }
+
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/ingested-entity-to-entity-mapper.ts]: Deleted

File: src/business-logic/services/ingested-entity-to-entity-mapper.ts [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+      playedDuration: partToBeUpdated.getPlayedDuration(),
+      endState: partToBeUpdated.getEndState(),
+      timings: this.getPartTimings(partToBeUpdated),
+      defaultPart: ingestedPart
+    })
+  }
+
+  private getPartTimings(part: Part): PartTimings | undefined {
+    try {
+      return  part.getTimings()
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/ingested-entity-to-entity-mapper.ts]: Removed

File: src/business-logic/services/ingested-entity-to-entity-mapper.ts [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+    } catch (error) {
+      if ((error as Exception).errorCode !== ErrorCode.UNSUPPORTED_OPERATION) {
+        throw error
+      }
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/ingested-entity-to-entity-mapper.ts]: Updated to your suggestion.
Your are right about your sidenote. It's been updated :)

File: src/business-logic/services/ingested-entity-to-entity-mapper.ts [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+      inTransition: ingestedPart.inTransition,
+      outTransition: ingestedPart.outTransition,
+      autoNext: ingestedPart.autoNext,
+      disableNextInTransition: ingestedPart.disableNextInTransition,
+      executedAt: partToBeUpdated.getExecutedAt(),
+      playedDuration: partToBeUpdated.getPlayedDuration(),
+      endState: partToBeUpdated.getEndState(),
+      timings: this.getPartTimings(partToBeUpdated),
+      defaultPart: ingestedPart
+    })
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/ingested-entity-to-entity-mapper.ts]: Updated

File: src/data-access/repositories/cache/cached-segment-repository.ts [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    const segmentIdsToDelete: string[] = []
+    this.cachedSegments.forEach((segment, segmentId) => {
+      if (predicate(segment)) {
+        segmentIdsToDelete.push(segmentId)
+      }
+    })
+    segmentIdsToDelete.forEach(id => this.cachedSegments.delete(id))
+  }
+
+  public async deleteAllUnsyncedSegments(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/cache/cached-segment-repository.ts:56]: Yes and no.
Technically it isn't, but I wanted to emphasize that _all_ unsynced Segments will be deleted by this and it isn't a method to call lightly

File: src/model/entities/part.ts [hunk 19]
```diff
@@ -302,13 +317,34 @@ export class Part {
...
   }
 
+  public getInTransition(): InTransition {
+    return this.inTransition
+  }
+
   public reset(): void {
+    if (!this.defaultPart) {
+      return
+    }
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/part.ts:322]: Nothing (except this) is stopping us from calling `reset` on an unplanned Part. Since an unplanned don't have a "default", we can't really reset them. (Besides, they are supposed to be removed when resetting the Segment).

So yes, it's technically possible for us to reach this point without a "defaultPart".

File: src/data-access/repositories/mongo/mongo-ingested-part-repository.ts [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    const mongoParts: MongoIngestedPart[] = (await this.getCollection()
+      .find<MongoIngestedPart>({ segmentId: segmentId })
+      .toArray())
+    const parts: IngestedPart[] = this.mongoIngestedEntityConverter.convertIngestedParts(mongoParts)
+    return Promise.all(
+      parts.map(async (ingestedPart) => {
+        ingestedPart.ingestedPieces = (await this.ingestedPieceRepository.getIngestedPieces(ingestedPart.id))
+        return ingestedPart
+      })
+    )
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-part-repository.ts:49]: In this particular instance changing it to your suggestion seems to be mainly semanctics (I am aware of the underlying difference).

I'm going to leave it as is for now. Your alternative suggestion would be a good idea if we need to improve performance on this.

File: src/data-access/repositories/mongo/mongo-ingested-part-repository.ts [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+    })
+    if (!mongoIngestedPart) {
+      throw new NotFoundException(`No Part found for partId: ${partId}`)
+    }
+    const ingestedPart: IngestedPart = this.mongoIngestedEntityConverter.convertIngestedPart(mongoIngestedPart)
+    ingestedPart.ingestedPieces = await this.ingestedPieceRepository.getIngestedPieces(ingestedPart.id)
+    return ingestedPart
+  }
+
+  public async getIngestedParts(segmentId: string): Promise<IngestedPart[]> {
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-part-repository.ts:38]: This comes down to your naming philosofy. 
One philosofy is to have everything in the name. Another philosofy is to have the parameter as part of the "name".
So by reading the parameter in conjunction with the method name, you would get that this method "get ingested parts for a segment".

As of the current implementation, the id of the IngestedSegment is the same as the id of Segment. Furthermore, the attribute is currently called `IngestedPart.segmentId`. Because of these reasons, I am going to leave it as is for now.

File: src/data-access/repositories/mongo/mongo-ingested-rundown-changed-listener.ts [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+import { IngestedRundownRepository } from '../interfaces/ingested-rundown-repository'
+import { IngestedRundown } from '../../../model/entities/ingested-rundown'
+
+const INGESTED_RUNDOWN_COLLECTION_NAME: string = 'rundowns' // TODO: Once we control ingest changed this to "ingestedRundowns"
+
+export class MongoIngestedRundownChangedListener extends BaseMongoRepository implements DataChangedListener<IngestedRundown> {
+
+  private onCreatedCallback: (rundown: IngestedRundown) => void
+  private onUpdatedCallback: (rundown: IngestedRundown) => void
+  private onDeletedCallback: (rundownId: string) => void
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-rundown-changed-listener.ts:23]: As of current implementation, yes

File: src/data-access/repositories/mongo/mongo-ingested-piece-repository.ts [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+    this.assertDatabaseConnection(this.getIngestedPieces.name)
+    const mongoPieces: MongoIngestedPiece[] = (await this.getCollection()
+      .find<MongoIngestedPiece>({ startPartId: partId })
+      .toArray())
+    return this.mongoEntityConverter.convertIngestedPieces(mongoPieces)
+  }
+
+  public async deleteIngestedPiecesForRundown(rundownId: string): Promise<void> {
+    this.assertDatabaseConnection(this.deleteIngestedPiecesForRundown.name)
+    await this.getCollection().deleteMany({ startRundownId: rundownId})
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-piece-repository.ts]: Updated

File: src/data-access/repositories/mongo/mongo-ingested-part-repository.ts [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+    this.assertDatabaseConnection(this.getIngestedParts.name)
+    const mongoParts: MongoIngestedPart[] = (await this.getCollection()
+      .find<MongoIngestedPart>({ segmentId: segmentId })
+      .toArray())
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-part-repository.ts]: No reason. They have been removed.

File: src/model/entities/ingested-part.ts [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+
+  inTransition: InTransition
+  outTransition: OutTransition
+
+  autoNext?: AutoNext
+  disableNextInTransition: boolean
+
+  isUntimed: boolean
+  timings?: PartTimings
+}
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/ingested-part.ts]: Made the variables immutable with the exception of `IngestedPart.ingestedPieces`, `IngestedSegment.ingestedParts`, `IngestedRundown.ingestedSegments` and `IngestedRundown.baselineTimelineObjects` with the reason being that the data for these attributes are fetched from different Collections in the database witch need to add to the interface somehow.

File: src/model/entities/piece.ts [hunk 26]
```diff
@@ -68,6 +69,14 @@ export class Piece {
     this.setExecutedAt(piece.executedAt ?? 0)
   }
 
+  public resetFromIngestedPiece(ingestedPiece: IngestedPiece): void {
+    this.start = ingestedPiece.start
+    if (this.pieceLifespan === PieceLifespan.WITHIN_PART) {
+      // Infinite Pieces might still be OnAir when their Part is resat, so we can't reset their "executedAt" here.
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/piece.ts]: Updated

File: src/blueprints/tv2/action-factories/tv2-camera-action-factory.ts [hunk 27]
```diff
@@ -100,6 +100,7 @@ export class Tv2CameraActionFactory {
   private createPartInterface(partId: string, source: Tv2SourceMappingWithSound): PartInterface {
     return {
       id: partId,
+      rundownId: '',
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/action-factories/tv2-camera-action-factory.ts:103]: I agree that it feels wrong.
I also agree that this isn't ideal. Don't have another solution yet, but when we move everything from blueprints to here, we should find one.

File: src/business-logic/services/database-change-ingest-service.ts (hunk 6 above)
Comment (LindvedKrvang) [reply src/business-logic/services/database-change-ingest-service.ts:107]: I considered it, but decided not because I don't think we gain that much from it.

File: src/business-logic/services/database-change-ingest-service.ts (hunk 7 above)
Comment (LindvedKrvang) [reply src/business-logic/services/database-change-ingest-service.ts]: Technically no, but I added it because I wanted to emphasize that we are creating a brand new Segment here.

File: src/business-logic/services/database-change-ingest-service.ts (hunk 10 above)
Comment (LindvedKrvang) [reply src/business-logic/services/database-change-ingest-service.ts]: Updated the names.
I'm more than open for alternatives for how to handle this, but it's going to be in another PR :)

File: src/business-logic/services/database-change-ingest-service.ts (hunk 9 above)
Comment (LindvedKrvang) [reply src/business-logic/services/database-change-ingest-service.ts]: Updated

File: src/business-logic/services/database-change-ingest-service.ts [hunk 28]
```diff
@@ -40,46 +55,170 @@ export class DatabaseChangeIngestService implements IngestChangeService {
...
+        this.eventEmitter.emitRundownDeleted(rundownId)
+      }
+    }
+  }
+
+  private updateActiveRundownFromIngestedRundown(ingestedRundown: IngestedRundown, rundown: Rundown): Rundown {
+    const ingestedSegmentIds: string[] = ingestedRundown.ingestedSegments.map(ingestedSegment => ingestedSegment.id)
+    const segmentsToBeDeleted: Segment[] = rundown.getSegments().filter(segment => !ingestedSegmentIds.includes(segment.id))
+    segmentsToBeDeleted.forEach(segment => rundown.removeSegment(segment.id))
+
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/database-change-ingest-service.ts:149]: Deleted

File: src/business-logic/services/database-change-ingest-service.ts [hunk 29]
```diff
@@ -40,46 +55,170 @@ export class DatabaseChangeIngestService implements IngestChangeService {
...
+        oldRundown = await this.rundownRepository.getRundown(ingestedRundown.id)
+      } catch (exception) {
+        if (!(exception instanceof NotFoundException)) {
+          throw exception
+        }
+        // The Rundown doesn't exist in the database which means it's a brand new Rundown.
+      }
+
+      // If the Rundown isn't active, we can simply just "re-ingest" it into our database collection as a fresh Rundown.
+      const updatedRundown: Rundown = oldRundown && oldRundown.isActive()
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/database-change-ingest-service.ts]: Updated

File: src/model/entities/rundown.ts [hunk 30]
```diff
@@ -66,10 +66,11 @@ export class Rundown extends BasicRundown {
 
   constructor(rundown: RundownInterface) {
     super(rundown.id, rundown.name, rundown.isRundownActive, rundown.modifiedAt, rundown.timing)
-    this.segments = rundown.segments ?? []
+    this.segments = rundown.segments ? [...rundown.segments].flat().sort(this.compareSegments) : []
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/rundown.ts:69]: It does. Without the `.flat` the returned type was `Segment[][]`

File: src/business-logic/services/database-change-ingest-service.ts [hunk 31]
```diff
@@ -40,46 +55,183 @@ export class DatabaseChangeIngestService implements IngestChangeService {
...
+
+    for (const basicRundown of basicRundowns) {
+      if (!ingestedRundownsIds.includes(basicRundown.id)) {
+        await this.rundownRepository.deleteRundown(basicRundown.id)
+        this.eventEmitter.emitRundownDeleted(basicRundown.id)
+      }
+    }
+  }
+
+  private async fetchOldRundown(rundownId: string): Promise<Rundown | undefined> {
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/database-change-ingest-service.ts:134]: I guess we should just call it `fetchRundown` since that is what it does?

[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/database-change-ingest-service.ts]: Updated

File: src/data-access/repositories/mongo/mongo-ingested-rundown-repository.ts [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+
+  private async populateIngestedRundown(mongoIngestedRundown: MongoIngestedRundown): Promise<IngestedRundown> {
+    const ingestedRundown: IngestedRundown = this.mongoIngestedEntityConverter.convertToIngestedRundown(mongoIngestedRundown)
+
+    ingestedRundown.ingestedSegments = await this.ingestedSegmentRepository.getIngestedSegments(ingestedRundown.id)
+    ingestedRundown.baselineTimelineObjects = await this.rundownBaselineRepository.getRundownBaseline(
+      ingestedRundown.id
+    )
+    return ingestedRundown
+  }
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-rundown-repository.ts]: Updated and changed the propteries to `readonly` on the interface.

File: src/data-access/repositories/mongo/mongo-ad-lib-actions-repository.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (andr9528) [reply src/data-access/repositories/mongo/mongo-ad-lib-actions-repository.ts:12]: Where is the new line?

File: src/model/entities/ingested-part.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/ingested-part.ts]: The interfaces are now completely readonly

File: src/model/entities/part.ts [hunk 33]
```diff
@@ -69,12 +74,15 @@ export class Part {
    */
   private endState?: PartEndState
 
+  public readonly defaultPart?: IngestedPart
```
Comment (andr9528) [inline src/model/entities/part.ts:77]: Seems that the `defaultPart` on the interface has been renamed to `ingestedPart`, unless it has always been named that. 
Should the name of the property on the class be the same name too?

**Review by KvelaGorrrrnio — COMMENTED** [review]
There are a few unresolved comments in the original comment list as well.

File: src/data-access/repositories/mongo/mongo-ingested-rundown-repository.ts (hunk 32 above)
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-rundown-repository.ts]: Updated

File: src/model/entities/rundown.ts [hunk 34]
```diff
@@ -546,19 +546,32 @@ export class Rundown extends BasicRundown {
     this.updateNextCursor()
   }
 
-  public removeSegment(segmentId: string): void {
-    const segment: Segment | undefined = this.segments.find(s => s.id === segmentId)
-    if (!segment) {
+  public removeUnsyncedSegment(unsyncedSegment: Segment): void {
+    if (unsyncedSegment.isOnAir()) {
+      throw new UnsupportedOperation('Trying to remove an unsynced from the Rundown while it is still on Air')
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/rundown.ts]: Updated

[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/model/entities/rundown.ts]: Not going do much more for this in this PR

File: src/data-access/repositories/mongo/mongo-ingested-part-repository.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-part-repository.ts]: It's been updated

File: src/model/entities/part.ts (hunk 33 above)
Comment (LindvedKrvang) [reply src/model/entities/part.ts]: Updated

File: src/data-access/repositories/mongo/mongo-ingested-rundown-repository.ts [hunk 35]
```diff
@@ -0,0 +1,63 @@
...
+    const mongoIngestedRundowns: MongoIngestedRundown[] = await this.getCollection().find<MongoIngestedRundown>({}).toArray()
+    return Promise.all(mongoIngestedRundowns.map(mongoIngestedRundown => this.populateIngestedRundown(mongoIngestedRundown)))
+  }
+
+  private async populateIngestedRundown(mongoIngestedRundown: MongoIngestedRundown): Promise<IngestedRundown> {
+    return {
+      ...this.mongoIngestedEntityConverter.convertToIngestedRundown(mongoIngestedRundown),
+      ingestedSegments: await this.ingestedSegmentRepository.getIngestedSegmentsByRundown(mongoIngestedRundown._id),
+      baselineTimelineObjects: await this.rundownBaselineRepository.getRundownBaseline(mongoIngestedRundown._id)
+
```
Comment (andr9528) [inline src/data-access/repositories/mongo/mongo-ingested-rundown-repository.ts:38]: Extra unneeded line.

**Review by andr9528 — APPROVED** [review]

File: src/data-access/repositories/mongo/mongo-ingested-part-repository.ts [hunk 36]
```diff
@@ -32,12 +32,12 @@ export class MongoIngestedPartRepository extends BaseMongoRepository implements
...
     return {
       ...this.mongoIngestedEntityConverter.convertToIngestedPart(mongoIngestedPart),
-      ingestedPieces: await this.ingestedPieceRepository.getIngestedPieces(mongoIngestedPart._id)
+      ingestedPieces: await this.ingestedPieceRepository.getIngestedPiecesByPart(mongoIngestedPart._id)
     }
   }
 
-  public async getIngestedParts(segmentId: string): Promise<IngestedPart[]> {
-    this.assertDatabaseConnection(this.getIngestedParts.name)
+  public async getIngestedPartsBySegment(segmentId: string): Promise<IngestedPart[]> {
```
Comment (KvelaGorrrrnio) [inline src/data-access/repositories/mongo/mongo-ingested-part-repository.ts:39]: I am glad for the change to add the scope of the request, but is there a difference between `BySegment` and `ForSegment`? We use currently use both.
It would be nice to only use one or the another (unless the meaning differs).

File: src/data-access/repositories/mongo/mongo-ingested-segment-changed-listener.ts [hunk 37]
```diff
@@ -42,7 +42,7 @@ export class MongoIngestedSegmentChangedListener extends BaseMongoRepository imp
       case MongoChangeEvent.INSERT: {
         const insertChange: ChangeStreamInsertDocument<MongoIngestedSegment> = change as ChangeStreamInsertDocument<MongoIngestedSegment>
         const ingestedSegmentId: string = insertChange.fullDocument._id
-        const ingestedSegment: IngestedSegment = await this.ingestedSegmentRepository.getIngestedSegment(ingestedSegmentId)
+        const ingestedSegment: IngestedSegment = await this.ingestedSegmentRepository.getIngestedSegmentRundown(ingestedSegmentId)
```
Comment (KvelaGorrrrnio) [inline src/data-access/repositories/mongo/mongo-ingested-segment-changed-listener.ts:45]: What is a Ingested Segment Rundown? 😁

**Review by KvelaGorrrrnio — APPROVED** [review]
I see good stuff! A typo and a question is all there is left 🚀

File: src/data-access/repositories/mongo/mongo-ingested-segment-changed-listener.ts (hunk 37 above)
Comment (LindvedKrvang) [reply src/data-access/repositories/mongo/mongo-ingested-segment-changed-listener.ts]: A spelling mistake :D 
It's been updated
